provider/github/watcher.go

```diff
@@ -73,34 +81,94 @@ func NewWatcher(transport http.RoundTripper, o *lookout.WatchOptions) (*Watcher,
 func (w *Watcher) Watch(ctx context.Context, cb lookout.EventHandler) error {
 	log.With(log.Fields{"urls": w.o.URLs}).Infof("Starting watcher")
 
-	repoIndex := 0
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	errCh := make(chan error)
+
+	for _, repo := range w.r {
+		go w.watchPrs(ctx, repo, cb, errCh)
+		go w.watchEvents(ctx, repo, cb, errCh)
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case err := <-errCh:
+		return err
+	}
+}
 
+func (w *Watcher) watchPrs(ctx context.Context, repo *lookout.RepositoryInfo, cb lookout.EventHandler, errCh chan error) {
 	for {
-		repo := w.r[repoIndex]
-		repoIndex = (repoIndex + 1) % len(w.r)
+		resp, prs, err := w.doPRListRequest(ctx, repo.Username, repo.Name)
+		if err != nil && !NoErrNotModified.Is(err) {
+			errCh <- err
+			return
+		}
+
+		if err := w.handlePrs(cb, repo, resp, prs); err != nil {
+			if lookout.NoErrStopWatcher.Is(err) {
+				return
+			}
 
+			errCh <- err
+			return
+		}
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(w.prPullInterval):
+			continue
+		}
+	}
+}
+
+func (w *Watcher) watchEvents(ctx context.Context, repo *lookout.RepositoryInfo, cb lookout.EventHandler, errCh chan error) {
+	for {
 		resp, events, err := w.doEventRequest(ctx, repo.Username, repo.Name)
 		if err != nil && !NoErrNotModified.Is(err) {
-			return err
+			errCh <- err
+			return
 		}
 
 		if err := w.handleEvents(cb, repo, resp, events); err != nil {
 			if lookout.NoErrStopWatcher.Is(err) {
-				return nil
+				return
 			}
 
-			return err
+			errCh <- err
+			return
 		}
 
 		select {
 		case <-ctx.Done():
-			return ctx.Err()
-		case <-time.After(w.pollInterval):
+			return
+		case <-time.After(w.eventsPollInterval):
```

Comment:
:thinking: I believe this will not obey the `X-Poll-Interval` header. `watchEvents` is run as a goroutine for each repo, and this `for` will wait the interval between each call _for the same repo_. But calls to different repos are actually done concurrently, without any interval between them.

Reply:
Hm... I don't see in the documentation if the `X-Poll-Interval` is for any `/events` or per repo. When I was writing this code I assumed it's per repo. Though now I think you are right.